logic/transactions/pool.js

```diff
@@ -0,0 +1,922 @@
+	transaction.receivedAt = new Date();
+	async.waterfall([
+		async.apply(__private.checkPoolAvailability, transaction),
+		__private.setAccountAndGet,
+		__private.getRequester,
+		__private.processTransaction,
+		__private.verifyTransaction,
+		__private.verifyTransactionTypeInPool,
+		function checkBalance (transaction, sender, waterCb) {
+			self.checkBalance(transaction, sender, function (err, balance) {
+				if (err) {
+					return setImmediate(waterCb, err);
+				}
+				return __private.moveToVerified(transaction, waterCb);
+			});
+		}
+	], function (err) {
+		return setImmediate(cb, err);
+	});
+
+};
+
+/**
+ * Adds transactions to unverified pool list.
+ * Checks if transaction is in pool, pool limit and performs basic transaction validations.
+ * @private
+ * @implements {__private.checkPoolAvailability}
+ * @implements {__private.setAccountAndGet}
+ * @implements {__private.getAccount}
+ * @implements {__private.processTransaction}
+ * @param {Transaction} transaction
+ * @param {boolean} broadcast
+ * @param {function} cb - Callback function.
+ * @return {setImmediateCallback} error | cb
+ */
+__private.addToUnverified = function (transaction, broadcast, cb) {
+	async.waterfall([
+		async.apply(__private.checkPoolAvailability, transaction),
+		__private.setAccountAndGet,
+		__private.getRequester,
+		__private.processTransaction
+	], function (err, transaction, sender) {
+		if (!err) {
+			if (broadcast) {
+				transaction.broadcast = true;
+			}
+			transaction.receivedAt = new Date();
+			pool.unverified.transactions[transaction.id] = transaction;
+			pool.unverified.count++;
+		}
+		return setImmediate(cb, err, transaction);
+	});
+
+};
+
+/**
+ * Adds transactions to pool list, increment counter only if transaction id is not present.
+ * @private
+ * @param {transaction} transaction
+ * @param {Object} poolList
+ * @param {function} cb - Callback function.
+ * @return {setImmediateCallback} error | cb
+ */
+__private.addToPoolList = function (transaction, poolList, cb) {
+	if (poolList.transactions[transaction.id] === undefined) {
+		poolList.count++;
+	}
+	poolList.transactions[transaction.id] = transaction;
+	return setImmediate(cb);
+};
+
+/**
+ * Adds transactions to pool list without checks.
+ * @private
+ * @param {transaction} transaction
+ * @param {function} cb - Callback function.
+ * @return {setImmediateCallback} error | cb
+ */
+__private.addReadyCheckBroadcast = function (transaction, cb) {
+	if (transaction.broadcast) {
+		delete transaction.broadcast;
+		pool.broadcast.push(transaction);
+	}
+	__private.addToPoolList(transaction, pool.verified.ready, cb);
+};
+
+/**
+ * Creates signature based on multisignature transaction and secret.
+ * @private
+ * @param {transaction} transaction
+ * @param {String} secret
+ * @param {function} cb - Callback function.
+ * @return {setImmediateCallback} error | cb
+ */
+__private.createSignature = function (transaction, secret, cb) {
```

Comment:
Should that be part of tx pool?

Reply:
Next version of lisk-js will send transactionId, publicKey and signature, pool will receive directly and if transaction id is in pending list, it will check the signature and add them to multisignature